<a href="https://colab.research.google.com/github/vengaladasvivek140-sudo/AIAC/blob/main/lab3_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

TASK1:Write a python program to Try 3 different prompts to generate a factorial function

In [1]:
# Prompt 1: Using a recursive function
def factorial_recursive(n):
  if n == 0:
    return 1
  else:
    return n * factorial_recursive(n-1)

print(f"Factorial of 5 (recursive): {factorial_recursive(5)}")

# Prompt 2: Using an iterative approach with a for loop
def factorial_iterative_for(n):
  result = 1
  for i in range(1, n + 1):
    result *= i
  return result

print(f"Factorial of 5 (iterative with for loop): {factorial_iterative_for(5)}")

# Prompt 3: Using an iterative approach with a while loop
def factorial_iterative_while(n):
  result = 1
  i = 1
  while i <= n:
    result *= i
    i += 1
  return result

print(f"Factorial of 5 (iterative with while loop): {factorial_iterative_while(5)}")

Factorial of 5 (recursive): 120
Factorial of 5 (iterative with for loop): 120
Factorial of 5 (iterative with while loop): 120


# Prompt 1: Using a recursive function
def factorial_recursive(n):
  if n == 0:
    return 1
  else:
    return n * factorial_recursive(n-1)

print(f"Factorial of 5 (recursive): {factorial_recursive(5)}")

# Prompt 2: Using an iterative approach with a for loop
def factorial_iterative_for(n):
  result = 1
  for i in range(1, n + 1):
    result *= i
  return result

print(f"Factorial of 5 (iterative with for loop): {factorial_iterative_for(5)}")

# Prompt 3: Using an iterative approach with a while loop
def factorial_iterative_while(n):
  result = 1
  i = 1
  while i <= n:
    result *= i
    i += 1
  return result

print(f"Factorial of 5 (iterative with while loop): {factorial_iterative_while(5)}")

EXPLAINATION:

This code demonstrates three different ways to calculate the factorial of a number (in this case, 5):

Recursive function (factorial_recursive): This function calls itself with a smaller argument until it reaches the base case (n=0), where it returns 1. The result is then calculated by multiplying the current number by the factorial of the previous number.
Iterative approach with a for loop (factorial_iterative_for): This function initializes a result variable to 1 and then iterates from 1 up to the given number (n), multiplying the result by each number in the range.
Iterative approach with a while loop (factorial_iterative_while): Similar to the for loop approach, this function initializes a result to 1 and a counter i to 1. It then uses a while loop that continues as long as i is less than or equal to n, multiplying the result by i and incrementing i in each iteration.
All three methods achieve the same result (120 for the factorial of 5) but use different programming techniques.

TASK2:Provide a clear example input-output prompt to generate a sorting function

In [5]:
# Prompt: Write a Python function called 'sort_list' that takes a list of numbers as input and returns a new list with the numbers sorted in ascending order.

# Example Input:
# input_list = [3, 1, 4, 1, 5, 9, 2, 6]

# Expected Output:
# output_list = [1, 1, 2, 3, 4, 5, 6, 9]

def sort_list(input_list):
  # Write your code here to sort the input_list and return the sorted list
  pass # Replace this pass statement with your sorting logic

# Test the function with the example input
# sorted_example_list = sort_list(input_list)
# print(sorted_example_list)

# Prompt: Write a Python function called 'sort_list' that takes a list of numbers as input and returns a new list with the numbers sorted in ascending order.

# Example Input:
# input_list = [3, 1, 4, 1, 5, 9, 2, 6]

# Expected Output:
# output_list = [1, 1, 2, 3, 4, 5, 6, 9]

def sort_list(input_list):
  # Write your code here to sort the input_list and return the sorted list
  pass # Replace this pass statement with your sorting logic

# Test the function with the example input
# sorted_example_list = sort_list(input_list)
# print(sorted_example_list)

EXPLAINATION:
This code snippet provides a template for creating a Python function called sort_list.

Here's a breakdown:

Prompt: This describes the goal of the function – to take a list of numbers and return a new list sorted in ascending order.
Example Input: This shows a sample list of numbers that would be used as input to the sort_list function.
Expected Output: This shows what the sort_list function should return when given the Example Input.
def sort_list(input_list):: This defines the function sort_list which takes one argument, input_list.
pass: This is a placeholder. You would replace this line with the actual code to sort the input_list.
Commented-out test code: The lines starting with # are commented out and show how you would typically test the function with the example input and print the result.
This template is designed to guide you in writing the sorting logic within the sort_list function. You would need to replace the pass statement with an algorithm that sorts the input_list and returns the sorted result.

Task3:Write a python program to Start with the vague prompt “Generate python code to calculate power bill” and
improve it step-by-step

In [4]:
# Prompt: Generate python code to calculate power bill

def calculate_power_bill():
  # This is a very basic placeholder based on the vague prompt.
  # We need more information to make it useful.
  print("This function is a placeholder for power bill calculation.")
  print("Please provide more details about how power bills are calculated (e.g., consumption, rates).")
  return 0 # Returning 0 as a placeholder result

# Example usage (will not do much yet)
bill_amount = calculate_power_bill()
print(f"Calculated bill amount: {bill_amount}")

This function is a placeholder for power bill calculation.
Please provide more details about how power bills are calculated (e.g., consumption, rates).
Calculated bill amount: 0


#EXPLAINATION:

This code is a very basic starting point for calculating a power bill, generated from the initial vague prompt.

# Prompt: Generate python code to calculate power bill: This is the original prompt used to generate the code.
def calculate_power_bill():: This defines a function named calculate_power_bill.
The comments within the function explain that this is a placeholder and more details are needed.
print("This function is a placeholder for power bill calculation.") and print("Please provide more details about how power bills are calculated (e.g., consumption, rates)."): These lines print messages indicating that the function is incomplete and requires more information.
return 0: This line returns 0 as a temporary placeholder value for the bill amount.
The lines outside the function show a basic example of how you would call the function and print the result.
Essentially, this code acknowledges the request but highlights that the initial prompt was too vague to create a functional power bill calculator. The next steps will involve refining the prompt to provide the necessary details for a more accurate calculation.

TASK4:Write a python program Write structured comments to help AI generate two linked functions (e.g., login_user()
and register_user())

In [3]:
# This module handles user authentication.

# Function: register_user
# Description: Registers a new user with a username and password.
# Inputs:
#   - username (str): The desired username.
#   - password (str): The user's password.
# Behavior:
#   - Checks if the username already exists.
#   - If not, securely stores the username and a hashed password.
#   - Returns True on successful registration, False otherwise.
def register_user(username, password):
  pass # Placeholder for registration logic

# Function: login_user
# Description: Logs in an existing user with a username and password.
# Inputs:
#   - username (str): The username.
#   - password (str): The user's password.
# Behavior:
#   - Checks if the username exists.
#   - If it exists, verifies the provided password against the stored hashed password.
#   - Returns True on successful login, False otherwise.
# Linked function: register_user (Users must be registered before logging in)
def login_user(username, password):
  pass # Placeholder for login logic

EXPLAINATION:

This code snippet uses structured comments to describe two linked Python functions: register_user and login_user.

# This module handles user authentication.: This is a general comment indicating the purpose of this code section.
# Function: register_user: This section of comments describes the register_user function.
# Description: Explains what the function does (registers a new user).
# Inputs: Specifies the expected inputs (username and password) and their data types (str).
# Behavior: Outlines the steps the function should perform (check for existing username, store credentials securely, return True/False based on success).
def register_user(username, password): pass: This is the actual function definition with a pass statement as a placeholder. The actual registration logic would go here.
# Function: login_user: This section of comments describes the login_user function.
# Description: Explains what the function does (logs in an existing user).
# Inputs: Specifies the expected inputs (username and password) and their data types (str).
# Behavior: Outlines the steps the function should perform (check for existing username, verify password, return True/False based on success).
# Linked function: register_user: This is an important comment that indicates the relationship between login_user and register_user, highlighting that a user must be registered before they can log in.
def login_user(username, password): pass: This is the actual function definition with a pass statement as a placeholder. The actual login logic would go here.
These structured comments serve as instructions for an AI (or another developer) to understand the intended functionality and relationships of these functions, making it easier to generate or write the actual code for registration and login.



TASK5:Write a program in python analyzing Prompt Specificity: Improving Temperature Conversion Function with
Clear Instructions

In [ ]:
# Vague Prompt: Write a function to convert temperature.

def convert_temperature(temp):
  # This is a placeholder based on the vague prompt.
  # We need more specific instructions to implement the conversion.
  print("This function is a placeholder for temperature conversion.")
  print("Please specify the units to convert from and to (e.g., Celsius to Fahrenheit).")
  return None # Returning None as a placeholder result

# Example usage (will not do much yet)
converted_temp = convert_temperature(25)
print(f"Converted temperature: {converted_temp}")

Explaination:
This code snippet is a result of using the vague prompt "Write a function to convert temperature".

# Vague Prompt: Write a function to convert temperature.: This comment indicates the prompt used to generate the code.
def convert_temperature(temp):: This defines a function named convert_temperature that takes one argument, temp.
The comments and print statements inside the function (print("This function is a placeholder...") and print("Please specify the units...")) explain that the function is a placeholder and requires more specific instructions to perform an actual temperature conversion.
return None: The function returns None as it cannot perform a meaningful conversion without knowing the input and output units.
The lines outside the function show a basic example of how you would call the function and print the result, although in this case, it will just print the placeholder messages and None.
This code effectively demonstrates that a vague prompt leads to a non-functional placeholder, highlighting the importance of providing clear and specific instructions when generating code.
